File: tmdb-core/src/main/java/dev/reuss/tmdb/core/exception/TmdbExceptions.java
package dev.reuss.tmdb.core.exception;

/**
 * Factory methods for mapping TMDB response status values to SDK exceptions.
 */
public final class TmdbExceptions {

    private TmdbExceptions() {
    }

    /**
     * Creates the most specific API exception for a parsed TMDB error response.
     *
     * @param httpStatus HTTP status code
     * @param tmdbStatusCode TMDB status code
     * @param tmdbStatusMessage TMDB status message
     * @param responseBody raw response body
     * @return mapped API exception
     */
    public static TmdbApiException from(
            int httpStatus,
            int tmdbStatusCode,
            String tmdbStatusMessage,
            String responseBody
    ) {
        return switch (httpStatus) {
            case 401, 403 -> new TmdbUnauthorizedException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody
            );
            case 404 -> new TmdbNotFoundException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody
            );
            case 429 -> new TmdbRateLimitException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody
            );
            case 500, 501, 502, 503, 504 -> new TmdbServerException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody
            );
            default -> new TmdbApiException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody
            );
        };
    }

    /**
     * Creates the most specific API exception for an HTTP error without a parsed TMDB status code.
     *
     * @param httpStatus HTTP status code
     * @param message error message
     * @param responseBody raw response body
     * @return mapped API exception
     */
    public static TmdbApiException fromHttpStatus(
            int httpStatus,
            String message,
            String responseBody
    ) {
        return switch (httpStatus) {
            case 401, 403 -> new TmdbUnauthorizedException(
                    message,
                    httpStatus,
                    0,
                    responseBody
            );
            case 404 -> new TmdbNotFoundException(
                    message,
                    httpStatus,
                    0,
                    responseBody
            );
            case 429 -> new TmdbRateLimitException(
                    message,
                    httpStatus,
                    0,
                    responseBody
            );
            case 500, 501, 502, 503, 504 -> new TmdbServerException(
                    message,
                    httpStatus,
                    0,
                    responseBody
            );
            default -> new TmdbApiException(
                    message,
                    httpStatus,
                    0,
                    responseBody
            );
        };
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/core/exception/TmdbExceptions.kt
package dev.reuss.tmdb.core.exception

/**
 * Maps HTTP and TMDB status information to the public exception hierarchy.
 */
object TmdbExceptions {
    /**
     * Selects the most specific exception supported for [httpStatus].
     *
     * Authentication and authorization errors map to [TmdbUnauthorizedException],
     * `404` to [TmdbNotFoundException], `429` to [TmdbRateLimitException], and
     * `500` through `504` to [TmdbServerException]. Other statuses remain
     * [TmdbApiException].
     */
    @JvmStatic
    fun from(
        httpStatus: Int,
        tmdbStatusCode: Int,
        tmdbStatusMessage: String,
        responseBody: String,
    ): TmdbApiException =
        when (httpStatus) {
            401, 403 -> {
                TmdbUnauthorizedException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody,
                )
            }

            404 -> {
                TmdbNotFoundException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody,
                )
            }

            429 -> {
                TmdbRateLimitException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody,
                )
            }

            in 500..504 -> {
                TmdbServerException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody,
                )
            }

            else -> {
                TmdbApiException(
                    tmdbStatusMessage,
                    httpStatus,
                    tmdbStatusCode,
                    responseBody,
                )
            }
        }

    /**
     * Maps an HTTP error whose body did not yield a TMDB status code.
     */
    @JvmStatic
    fun fromHttpStatus(
        httpStatus: Int,
        message: String,
        responseBody: String,
    ): TmdbApiException =
        from(
            httpStatus = httpStatus,
            tmdbStatusCode = 0,
            tmdbStatusMessage = message,
            responseBody = responseBody,
        )
}
